src/sentry/services/hybrid_cloud/email/__init__.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import annotations
+
+import abc
+from dataclasses import dataclass
+from typing import Collection, Iterable, Type
+
+from sentry.services.hybrid_cloud import InterfaceWithLifecycle, silo_mode_delegation, stubbed
+from sentry.services.hybrid_cloud.organization import ApiOrganization, ApiOrganizationMember
+from sentry.services.hybrid_cloud.user import APIUser
+from sentry.silo import SiloMode
+from sentry.utils import metrics
+
+
+@dataclass
+class ApiUserEmail:
+    id: int
+    user: APIUser
+    email: str
+    is_verified: bool
+    is_primary: bool
+
+
+class AmbiguousUserFromEmail(Exception):
+    def __init__(self, email: str, users: Collection[APIUser]) -> None:
+        super().__init__(f"Resolved {email!r} to {[user.id for user in users]}")
+        self.email = email
+        self.users = tuple(users)
+
+
+class EmailService(InterfaceWithLifecycle):
+    def resolve_email_to_user(
+        self, email: str, organization: ApiOrganization | None = None
+    ) -> APIUser | None:
+        """Resolve an email address to a user profile.
+
+        If no users with the email address are found, return None. If multiple users
+        with the email address are found and the steps below cannot disambiguate from
+        among them, raise AmbiguousUserFromEmail.
+
+        The disambiguation steps are as follows:
+
+        1. If some email addresses are unverified, exclude them.
+        2. If an organization is passed as an argument, prefer users with a membership
+           in that organization.
+        3. If one user has the email address as the primary address on their account,
+           choose that one.
+
+        :param email: the email address to resolve
+        :param organization: the organization to prioritize, if any
+        :return: the user, if found
+        """
+
+        candidates = tuple(self.get_user_emails(email=email))
+        if not candidates:
+            return None
+        return _EmailResolver(self, email, organization).resolve(candidates)
+
+    @abc.abstractmethod
+    def get_user_emails(self, *, email: str) -> Iterable[ApiUserEmail]:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_members_for_users(
+        self, *, organization: ApiOrganization, users: Collection[APIUser]
+    ) -> Iterable[ApiOrganizationMember]:
+        raise NotImplementedError
```

Comment:
Resolved from sync discussion: move this to `OrganizationService`.

Reply:
`ApiOrganizationMember` will be unavailable in control silo. Would need to replace with minimal member mapping when it exists. Or, if we're interested only in existence of membership, just return a subset of the `users` arg.